### ISSUE TITLE
e2e TestGitPipelineRun test fails on newly installed pipeline

### ISSUE THREAD

**Issue by barthy1** [tektoncd/pipeline#3627]
# Expected Behavior
`TestGitPipelineRun` test should pass

# Actual Behavior
`TestGitPipelineRun` test fails because of timeout. Corresponding taskrun is created but no pod is started.
Usually second attempt is successful.

```
# oc get tr -A
NAMESPACE         NAME                                     SUCCEEDED   REASON   STARTTIME   COMPLETIONTIME
arendelle-8nxtw   git-check-pipeline-run-git-check-72mth

# oc get pod -n arendelle-8nxtw
No resources found in arendelle-8nxtw namespace.
```

# Steps to Reproduce the Problem

1. Install k8s cluster using `kind`
2. Install nightly pipeline build
```
kubectl apply --filename https://storage.googleapis.com/tekton-releases-nightly/pipeline/latest/release.yaml
```
3. Run the test just after pipeline is installed
```
# go test -v -tags=e2e -count=1 ./test -run ^TestGitPipelineRun/tekton_@_master_with_PR_refspec
```
4. Wait 10mins and get the error 
```
# go test -v -tags=e2e -count=1 ./test -run ^TestGitPipelineRun/tekton_@_master_with_PR_refspec
Using kubeconfig at `/root/.kube/config` with cluster ``
Cluster information
- Version: v1.19.1
- BuildDate: 2020-09-14T07:30:52Z
- GoVersion: go1.15
- Platform: linux/amd64
=== RUN   TestGitPipelineRun
=== RUN   TestGitPipelineRun/tekton_@_master_with_PR_refspec
=== PAUSE TestGitPipelineRun/tekton_@_master_with_PR_refspec
=== CONT  TestGitPipelineRun/tekton_@_master_with_PR_refspec
    init_test.go:147: Create namespace arendelle-8nxtw to deploy to
    init_test.go:163: Verify SA "default" is created in namespace "arendelle-8nxtw"
    git_checkout_test.go:103: Creating Git PipelineResource git-source-resource
    git_checkout_test.go:119: Creating PipelineRun git-check-pipeline-run
panic: test timed out after 10m0s
...
FAIL	github.com/tektoncd/pipeline/test	600.088s
FAIL
```
# Additional Info

- Kubernetes version:
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.0", GitCommit:"af46c47ce925f4c4ad5cc8d1fca46c7b77d13b38", GitTreeState:"clean", BuildDate:"2020-12-08T17:59:43Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-14T07:30:52Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

- Tekton Pipeline version:
```
# tkn version
Client version: 0.11.0
Pipeline version: v20201210-d58a1ae68f
Triggers version: unknown
```

/kind bug

**Comment by nikhil-thomas** [comment]
does `kubectl describe taskrun  git-check-pipeline-run-git-check-72mth -n arendelle-8nxtw ` give any additional information.

**Comment by chmouel** [comment]
You probably would want to set this environment variable to be able to inspect the logs of the failed container : 

https://github.com/tektoncd/pipeline/blob/d702ca1cef1d156887e9306dd3739690b0ceb599/test/init_test.go#L111-L116

**Comment by barthy1** [comment]
@nikhil-thomas don't have the taskrun mentioned in this issue any more. Providing describe for new taskrun. From my point of view no additional info there.
```
# oc describe taskrun git-check-pipeline-run-git-check-7m7kw -n arendelle-lf2x8
Name:         git-check-pipeline-run-git-check-7m7kw
Namespace:    arendelle-lf2x8
Labels:       app.kubernetes.io/managed-by=tekton-pipelines
              tekton.dev/pipeline=git-check-pipeline-run
              tekton.dev/pipelineRun=git-check-pipeline-run
              tekton.dev/pipelineTask=git-check
Annotations:  <none>
API Version:  tekton.dev/v1beta1
Kind:         TaskRun
Metadata:
  Creation Timestamp:  2020-12-14T09:40:42Z
  Generation:          1
  Managed Fields:
    API Version:  tekton.dev/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          .:
          f:tekton.dev/pipeline:
          f:tekton.dev/pipelineRun:
          f:tekton.dev/pipelineTask:
        f:ownerReferences:
          .:
          k:{"uid":"5a46cb1b-7c3e-4c58-9c0a-68fe9bd4942f"}:
            .:
            f:apiVersion:
            f:blockOwnerDeletion:
            f:controller:
            f:kind:
            f:name:
            f:uid:
      f:spec:
        .:
        f:resources:
          .:
          f:inputs:
        f:serviceAccountName:
        f:taskSpec:
          .:
          f:resources:
            .:
            f:inputs:
          f:steps:
        f:timeout:
      f:status:
        .:
        f:podName:
    Manager:    controller
    Operation:  Update
    Time:       2020-12-14T09:40:42Z
  Owner References:
    API Version:           tekton.dev/v1beta1
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  PipelineRun
    Name:                  git-check-pipeline-run
    UID:                   5a46cb1b-7c3e-4c58-9c0a-68fe9bd4942f
  Resource Version:        2042242
  Self Link:               /apis/tekton.dev/v1beta1/namespaces/arendelle-lf2x8/taskruns/git-check-pipeline-run-git-check-7m7kw
  UID:                     99d06efb-23d4-4730-8958-0c7692d81135
Spec:
  Resources:
    Inputs:
      Name:  gitsource
      Resource Ref:
        Name:            git-source-resource
  Service Account Name:  default
  Task Spec:
    Resources:
      Inputs:
        Name:  gitsource
        Type:  git
    Steps:
      Args:
        --git-dir=/workspace/gitsource/.git
        show
      Image:  alpine/git
      Name:
      Resources:
  Timeout:  1h0m0s
Events:
  Type    Reason   Age   From     Message
  ----    ------   ----  ----     -------
  Normal  Started  12m   TaskRun
```
@chmouel no pods are actually created, so cannot inspect the logs.

**Comment by barthy1** [comment]
Update from my side, in 18mins after starting the test, taskrun was finished successfully and it took just several seconds. But obviously the test failed because of 10mins waiting timeout. So I observe some kind of delay for taskrun execution, when I start it first time.

**Comment by afrittoli** [comment]
> Update from my side, in 18mins after starting the test, taskrun was finished successfully and it took just several seconds. But obviously the test failed because of 10mins waiting timeout. So I observe some kind of delay for taskrun execution, when I start it first time.

Do you wait for all pods to be ready in the `tekton-pipelines` namespace before starting to run the tests?

**Comment by afrittoli** [comment]
> > Update from my side, in 18mins after starting the test, taskrun was finished successfully and it took just several seconds. But obviously the test failed because of 10mins waiting timeout. So I observe some kind of delay for taskrun execution, when I start it first time.
> 
> Do you wait for all pods to be ready in the `tekton-pipelines` namespace before starting to run the tests?

Looking at the task that installs Tekton it looks like you do, I wonder if there is something else causing the delay... it would be useful if you could share the controller logs?

**Comment by barthy1** [comment]
> Do you wait for all pods to be ready in the tekton-pipelines namespace before starting to run the tests?

yes

> Looking at the task that installs Tekton it looks like you do, I wonder if there is something else causing the delay... it would be useful if you could share the controller logs?

sure, I will provide the logs when do clean setup

**Comment by barthy1** [comment]
the logs from controller from start till test started to execute 
https://gist.github.com/barthy1/58e7fe5bc5dd67335c0c24a93894ae83#file-tekton_controller_logs

**Comment by sbwsg** [comment]
@barthy1 does this remain an issue in more recent e2e test runs?

**Comment by barthy1** [comment]
@sbwsg 
>does this remain an issue in more recent e2e test runs?

Yes, I see the same behaviour with the latest main code

**Comment by sbwsg** [comment]
There's a very sudden jump in the logs from 17:20 to 17:37 where nothing appears to be happening, and then an error is reported related to aws credentials:

```
{"level":"info","ts":"2021-02-17T17:20:12.695Z","logger":"tekton-pipelines-controller","caller":"controller/controller.go:530","msg":"Reconcile succeeded","commit":"736e98e","knative.dev/controller":"github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.Reconciler","knative.dev/kind":"tekton.dev.PipelineRun","knative.dev/traceid":"eb0c2a9b-396c-481d-b7e5-1207486aea80","knative.dev/key":"arendelle-2rtch/task-run-pipeline-run-cancel-retries-1-pdnjuwjf","duration":0.000033238}
E0217 17:37:55.174255       1 aws_credentials.go:77] while getting AWS credentials NoCredentialProviders: no valid providers in chain. Deprecated.
	For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

That error comes from `k8schain` (more accurately from the [`k8s-pkg-credentialprovider` package](https://github.com/vdemeester/k8s-pkg-credentialprovider/blob/f3505df23e86a46c6986c3c5577e129c7fe2ced6/aws/aws_credentials.go#L138-L141)) and so I believe that this process is stuck waiting for the `k8schain` package to do ... something (maybe timeout initializing aws creds? but where did the 17 minute delay come from?).  I'm just not sure what is happening yet.  I've been seeing a similar problem locally in Kind working on a separate issue.

Pulling in some more folks here who I mentioned the delay to yesterday on Slack. fyi @vdemeester and @imjasonh .

**Comment by vdemeester** [comment]
This sounds heavily similar to the GCP error we have with OpenShift (where the URL that `k8s-pkg-credentialprovider` does query a magic url and waits forever for an answer, completely locking the controller). 
The fork is exists only to remove the dependency on `k8s.io/kubernetes` (which brings *way too much* to the dependency graph) and the usage of it is in [`google/go-containerregistry`](https://github.com/google/go-containerregistry). 

Something might have changed in the recent version of this package and/or on the service side of AWS 🤔

**Comment by sbwsg** [comment]
Using the `runtime.trace` package I'm able to isolate 8x 75s-long calls to `http.dialConnFor`, that look like they're all spawned from the `aws-sdk-go` package.  So at least in my case I think this accounts for the ten minute delay:

```
0.000342165	 .         	198	task k8schain (id 1, parent 0) created
0.000345207	 .     3042	198	namespace=default
0.000346755	 .     1548	198	saName=default
0.000348249	 .     1494	198	region k8schain.New started (duration: 10m2.20341841s)
0.005701440	 .  5353191	198	new goroutine 260: github.com/aws/aws-sdk-go/internal/sync/singleflight.(*Group).doCall
0.005805444	 .   104004	260	new goroutine 261: net/http.(*Transport).dialConnFor
75.329407376	75.323601932	260	new goroutine 464: net/http.(*Transport).dialConnFor
150.583420726	75.254013350	260	new goroutine 785: net/http.(*Transport).dialConnFor
225.953546323	75.370125597	260	new goroutine 991: net/http.(*Transport).dialConnFor
301.129792264	75.176245941	260	new goroutine 1258: net/http.(*Transport).dialConnFor
376.398534187	75.268741923	260	new goroutine 1497: net/http.(*Transport).dialConnFor
451.715118090	75.316583903	260	new goroutine 1809: net/http.(*Transport).dialConnFor
527.040350774	75.325232684	260	new goroutine 2069: net/http.(*Transport).dialConnFor
602.203773864	75.163423090	198	total-k8schain-time=10m2.203418063s
602.203775323	 .     1459	198	task end
GC:22.952305ms
```

**Comment by sbwsg** [comment]
Searching around for similar issues it does sound like the aws sdk introduced a security feature to their SDK that has caused lots of difficult-to-debug delays for other projects.

Here's one example with a good explanation of what's going on: https://github.com/zalando-incubator/kube-ingress-aws-controller/issues/309#issuecomment-605099408

**Comment by sbwsg** [comment]
One workaround that appears to speed things up is to set nonsense AWS credentials in environment variables.  I've updated the tekton controller deployment with the following change:

```diff
λ git diff
diff --git a/config/controller.yaml b/config/controller.yaml
index d07084c5d..728776b4c 100644
--- a/config/controller.yaml
+++ b/config/controller.yaml
@@ -89,6 +89,12 @@ spec:
         - name: config-registry-cert
           mountPath: /etc/config-registry-cert
         env:
+        - name: AWS_ACCESS_KEY_ID
+          value: foobarbaz
+        - name: AWS_SECRET_ACCESS_KEY
+          value: foobarbaz
+        - name: AWS_DEFAULT_REGION
+          value: the-event-horizon
```

@barthy1 could you try doing something similar with your pipelines deployment and see if that changes the slowness?

**Comment by barthy1** [comment]
@sbwsg after I added the suggested env variables, I do not observe the slowness mentioned in this issue any more 👍

**Comment by sbwsg** [comment]
Awesome, thanks for the update.  This is very concerning.  I wonder how many projects are embedding k8schain or otherwise initializing the aws sdk and therefore slowing everything down just because it's a transitive dependency.

**Comment by bobcatfish** [comment]
Does #4073 fix this?

**Comment by barthy1** [comment]
@bobcatfish yes, with the latest pipeline code I do not observe the problems mentioned in this issue.
